integration/app_integration_test.go

```diff
@@ -0,0 +1,430 @@
+}
+
+// createWebSession creates a Web UI session within the root cluster.
+func (p pack) createWebSession(t *testing.T) (string, string) {
+	csReq, err := json.Marshal(web.CreateSessionReq{
+		User: p.username,
+		Pass: p.password,
+	})
+	require.NoError(t, err)
+
+	// Create POST request to create session.
+	u := url.URL{
+		Scheme: "https",
+		Host:   net.JoinHostPort(Loopback, p.rootCluster.GetPortWeb()),
+		Path:   "/v1/webapi/sessions/web",
+	}
+	req, err := http.NewRequest(http.MethodPost, u.String(), bytes.NewBuffer(csReq))
+	require.NoError(t, err)
+
+	// Attack CSRF token in cookie and header.
+	csrfToken, err := utils.CryptoRandomHex(32)
+	require.NoError(t, err)
+	req.AddCookie(&http.Cookie{
+		Name:  csrf.CookieName,
+		Value: csrfToken,
+	})
+	req.Header.Set("Content-Type", "application/json; charset=utf-8")
+	req.Header.Set(csrf.HeaderName, csrfToken)
+
+	// Issue request.
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		},
+	}
+	resp, err := client.Do(req)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+	require.Equal(t, http.StatusOK, resp.StatusCode)
+
+	// Read in response.
+	var csResp *web.CreateSessionResponse
+	err = json.NewDecoder(resp.Body).Decode(&csResp)
+	require.NoError(t, err)
+
+	// Extract session cookie and bearer token.
+	require.Len(t, resp.Cookies(), 1)
+	cookie := resp.Cookies()[0]
+	require.Equal(t, cookie.Name, web.CookieName)
+
+	return cookie.Value, csResp.Token
+}
+
+// createAppSession creates an application session with the root cluster. The
+// application that the user connects to may be running in a leaf cluster.
+func (p pack) createAppSession(t *testing.T, sessionCookie string, bearerToken string, publicAddr, clusterName string) string {
+	casReq, err := json.Marshal(web.CreateAppSessionRequest{
+		FQDN:        publicAddr,
+		PublicAddr:  publicAddr,
+		ClusterName: clusterName,
+	})
+	require.NoError(t, err)
+
+	u := url.URL{
+		Scheme: "https",
+		Host:   net.JoinHostPort(Loopback, p.rootCluster.GetPortWeb()),
+		Path:   "/v1/webapi/sessions/app",
+	}
+	req, err := http.NewRequest(http.MethodPost, u.String(), bytes.NewBuffer(casReq))
+	require.NoError(t, err)
+
+	req.AddCookie(&http.Cookie{
+		Name:  web.CookieName,
+		Value: sessionCookie,
+	})
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %v", bearerToken))
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		},
+	}
+	resp, err := client.Do(req)
+	require.NoError(t, err)
+	defer resp.Body.Close()
+	require.Equal(t, http.StatusOK, resp.StatusCode)
+
+	var casResp *web.CreateAppSessionResponse
+	err = json.NewDecoder(resp.Body).Decode(&casResp)
+	require.NoError(t, err)
+
+	return casResp.CookieValue
+}
+
+// makeRequest makes a request to the root cluster with the given session cookie.
+func (p *pack) makeRequest(sessionCookie string) (int, string, error) {
```

Comment:
Named return args wouldn't hurt in this case, to help explain what is being returned.

Reply:
Not a huge fan of named return arguments, I feel they encourage bare returns which are also very confusing from a readers perspective. Will think about how to refactor this method as I add more tests to this package.